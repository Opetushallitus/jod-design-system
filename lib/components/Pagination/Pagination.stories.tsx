import { PaginationItemLabelDetails } from '@ark-ui/react';
import type { StoryObj } from '@storybook/react-vite';
import { useState } from 'storybook/preview-api';
import { fn } from 'storybook/test';
import type { TitledMeta } from '../../utils';
import { Pagination } from './Pagination';

const meta = {
  title: 'Content/Pagination',
  component: Pagination,
  tags: ['autodocs'],
  parameters: {
    controls: {
      exclude: ['translations', 'onPageChange'],
    },
  },
  argTypes: {
    serviceVariant: {
      control: { type: 'radio' },
      options: ['yksilo', 'ohjaaja', 'tietopalvelu', 'palveluportaali'],
    },
  },
} satisfies TitledMeta<typeof Pagination>;

export default meta;

type Story = StoryObj<typeof meta>;

const render = (args: Story['args']) => {
  const { currentPage, onPageChange, ...rest } = args;
  const [pageNumber, setPageNumber] = useState<number>(currentPage);
  return (
    <Pagination
      currentPage={pageNumber}
      onPageChange={(newPage) => {
        setPageNumber(newPage.page);
        onPageChange(newPage);
      }}
      {...rest}
    />
  );
};

const url = 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542%3A6678';
const translations = {
  prevTriggerLabel: 'Edellinen',
  nextTriggerLabel: 'Seuraava',
  itemLabel: (item: PaginationItemLabelDetails) => `Sivu ${item.page}`,
  firstPageTriggerLabel: 'Ensimm√§inen sivu',
  lastPageTriggerLabel: 'Viimeinen sivu',
};

export const Default: Story = {
  render,
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'Pagination component for navigating between pages.',
      },
    },
  },
  args: {
    totalItems: 30,
    pageSize: 20,
    currentPage: 1,
    siblingCount: 2,
    onPageChange: fn(),
    translations,
  },
};

export const Ellipsis: Story = {
  render,
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'Pagination when there are more pages than can be displayed.',
      },
    },
  },
  args: {
    totalItems: 5000,
    pageSize: 20,
    currentPage: 1,
    siblingCount: 2,
    onPageChange: fn(),
    translations,
  },
};
