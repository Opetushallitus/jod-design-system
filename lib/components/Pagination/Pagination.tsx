import { Pagination as ArkPagination, PaginationRootProps } from '@ark-ui/react';
import { cx } from '../../cva';
import { JodMore, JodPagerNext, JodPagerPrev } from '../../icons';
import { getAccentBgClassForService, ServiceVariant } from '../../utils';

interface GetClassNameProps {
  isActive?: boolean;
  isArrowButton?: boolean;
  disabled?: boolean;
  isEllipsis?: boolean;
  serviceVariant?: ServiceVariant;
}
const getClassName = ({
  isActive = false,
  isArrowButton = true,
  disabled = false,
  isEllipsis = false,
  serviceVariant = 'yksilo',
}: GetClassNameProps) =>
  cx(
    `ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center`,
    {
      'ds:cursor-pointer': !isEllipsis || !disabled,
      'ds:cursor-default': isEllipsis,
      'ds:disabled:text-inactive-gray ds:disabled:cursor-not-allowed': disabled === true,
      [`${getAccentBgClassForService(serviceVariant)} ds:text-white`]: !isArrowButton && isActive,
      'ds:bg-bg-gray-2 ds:text-primary-gray': !isActive,
      'ds:font-bold': !isArrowButton,
    },
  );

export interface PageChangeDetails {
  page: number;
  pageSize: number;
}

export interface PaginationProps {
  totalItems: number;
  pageSize: number;
  siblingCount: number;
  currentPage: number;
  translations: PaginationRootProps['translations'];
  onPageChange: (details: PageChangeDetails) => void;
  type?: 'button' | 'link';
  ariaLabel?: string;
  serviceVariant?: ServiceVariant;
  dataTestId?: string;
}

/** Pagination component for navigating through a list of items. */
export const Pagination = ({
  totalItems,
  pageSize,
  siblingCount,
  currentPage,
  translations,
  type,
  ariaLabel,
  onPageChange,
  serviceVariant = 'yksilo',
  dataTestId,
}: PaginationProps) => {
  const lastPage = Math.ceil(totalItems / pageSize);
  const isFirstPage = currentPage === 1;
  const isLastPage = currentPage === lastPage;

  return (
    <ArkPagination.Root
      count={totalItems}
      pageSize={pageSize}
      page={currentPage}
      siblingCount={siblingCount}
      translations={translations}
      onPageChange={onPageChange}
      type={type}
      aria-label={ariaLabel}
      className="ds:inline-flex ds:list-none ds:items-center ds:justify-center ds:gap-3"
      data-testid={dataTestId}
    >
      <ArkPagination.PrevTrigger className={getClassName({ disabled: isFirstPage })} disabled={isFirstPage}>
        <JodPagerPrev size={24} />
      </ArkPagination.PrevTrigger>
      <ArkPagination.Context>
        {(pagination) =>
          pagination.pages.map((page, index) =>
            page.type === 'page' ? (
              <ArkPagination.Item
                key={page.value}
                {...page}
                className={getClassName({ isArrowButton: false, isActive: currentPage === page.value, serviceVariant })}
              >
                {page.value}
              </ArkPagination.Item>
            ) : (
              <ArkPagination.Ellipsis
                // The page object for ellipsis does not contain a value, so index must be used as a key.
                // eslint-disable-next-line react/no-array-index-key
                key={`ellipsis_${index}`}
                index={index}
                className={getClassName({ isEllipsis: true })}
              >
                <JodMore size={24} className="ds:rotate-90" />
              </ArkPagination.Ellipsis>
            ),
          )
        }
      </ArkPagination.Context>
      <ArkPagination.NextTrigger className={getClassName({ disabled: isLastPage })} disabled={isLastPage}>
        <JodPagerNext size={24} />
      </ArkPagination.NextTrigger>
    </ArkPagination.Root>
  );
};
