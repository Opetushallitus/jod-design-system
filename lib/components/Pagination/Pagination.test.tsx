import { render, screen, waitFor } from '@testing-library/react';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { Pagination, PaginationProps } from './Pagination';

describe('Pagination', () => {
  const mockTranslations = {
    prevTriggerLabel: 'Previous',
    nextTriggerLabel: 'Next',
  };

  const mockOnPageChange = vi.fn();

  afterEach(() => {
    mockOnPageChange.mockReset();
  });

  const renderPagination = (props: PaginationProps) =>
    render(
      <Pagination
        totalItems={props.totalItems}
        pageSize={props.pageSize}
        siblingCount={props.siblingCount}
        currentPage={props.currentPage}
        translations={props.translations}
        onPageChange={mockOnPageChange}
      />,
    );

  it('renders the pagination component with correct number of pages without ellipsis', () => {
    const props: PaginationProps = {
      totalItems: 100,
      pageSize: 25,
      siblingCount: 4,
      currentPage: 1,
      translations: mockTranslations,
      onPageChange: mockOnPageChange,
    };

    const { container } = renderPagination(props);

    const pageElements = screen.getAllByRole('button', { name: /[0-9]+/ });
    expect(pageElements.length).toBe(4);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the pagination component with correct number of pages with ellipsis', () => {
    const props: PaginationProps = {
      totalItems: 100,
      pageSize: 10,
      siblingCount: 1,
      currentPage: 1,
      translations: mockTranslations,
      onPageChange: mockOnPageChange,
    };

    const { container } = renderPagination(props);

    const pageElements = screen.getAllByRole('button', { name: /[0-9]+/ });
    expect(pageElements.length).toBe(6);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls onPageChange when a page is clicked', async () => {
    const props: PaginationProps = {
      totalItems: 100,
      pageSize: 10,
      siblingCount: 1,
      currentPage: 1,
      translations: mockTranslations,
      onPageChange: mockOnPageChange,
    };

    renderPagination(props);

    const pageElement = screen.getByRole('button', { name: '2' });
    pageElement.click();
    await waitFor(() => expect(mockOnPageChange).toHaveBeenCalledWith({ page: 2, pageSize: 10 }));
  });

  it('goes to first page', async () => {
    const props: PaginationProps = {
      totalItems: 100,
      pageSize: 5,
      siblingCount: 1,
      currentPage: 5,
      translations: mockTranslations,
      onPageChange: mockOnPageChange,
    };

    renderPagination(props);

    const pageElement = screen.getByRole('button', { name: '1' });
    pageElement.click();
    await waitFor(() => expect(mockOnPageChange).toHaveBeenCalledWith({ page: 1, pageSize: 5 }));
  });

  it('goes to last page', async () => {
    const props: PaginationProps = {
      totalItems: 100,
      pageSize: 5,
      siblingCount: 1,
      currentPage: 1,
      translations: mockTranslations,
      onPageChange: mockOnPageChange,
    };

    renderPagination(props);

    const pageElement = screen.getByRole('button', { name: '20' });
    pageElement.click();
    await waitFor(() => expect(mockOnPageChange).toHaveBeenCalledWith({ page: 20, pageSize: 5 }));
  });
});
