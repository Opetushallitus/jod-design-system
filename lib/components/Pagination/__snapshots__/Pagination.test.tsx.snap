// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > renders the pagination component with correct number of pages with ellipsis 1`] = `
<nav
  class="ds:inline-flex ds:list-none ds:items-center ds:justify-center ds:gap-3"
  data-part="root"
  data-scope="pagination"
  dir="ltr"
  id="pagination::r1:"
>
  <button
    aria-label="Previous"
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:disabled:text-inactive-gray ds:disabled:cursor-not-allowed ds:bg-bg-gray-2 ds:text-black"
    data-disabled=""
    data-part="prev-trigger"
    data-scope="pagination"
    dir="ltr"
    disabled=""
    id="pagination::r1::prev"
    type="button"
  >
    <svg
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
      />
    </svg>
  </button>
  <button
    aria-current="page"
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-accent ds:text-white ds:font-bold"
    data-index="1"
    data-part="item"
    data-scope="pagination"
    data-selected=""
    dir="ltr"
    id="pagination::r1::item:1"
    type="button"
  >
    1
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="2"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::item:2"
    type="button"
  >
    2
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="3"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::item:3"
    type="button"
  >
    3
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="4"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::item:4"
    type="button"
  >
    4
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="5"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::item:5"
    type="button"
  >
    5
  </button>
  <div
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-default ds:bg-bg-gray-2 ds:text-black"
    data-part="ellipsis"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::ellipsis:5"
  >
    <svg
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </div>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="10"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::item:10"
    type="button"
  >
    10
  </button>
  <button
    aria-label="Next"
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black"
    data-part="next-trigger"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r1::next"
    type="button"
  >
    <svg
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </button>
</nav>
`;

exports[`Pagination > renders the pagination component with correct number of pages without ellipsis 1`] = `
<nav
  class="ds:inline-flex ds:list-none ds:items-center ds:justify-center ds:gap-3"
  data-part="root"
  data-scope="pagination"
  dir="ltr"
  id="pagination::r0:"
>
  <button
    aria-label="Previous"
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:disabled:text-inactive-gray ds:disabled:cursor-not-allowed ds:bg-bg-gray-2 ds:text-black"
    data-disabled=""
    data-part="prev-trigger"
    data-scope="pagination"
    dir="ltr"
    disabled=""
    id="pagination::r0::prev"
    type="button"
  >
    <svg
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12z"
      />
    </svg>
  </button>
  <button
    aria-current="page"
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-accent ds:text-white ds:font-bold"
    data-index="1"
    data-part="item"
    data-scope="pagination"
    data-selected=""
    dir="ltr"
    id="pagination::r0::item:1"
    type="button"
  >
    1
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="2"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r0::item:2"
    type="button"
  >
    2
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="3"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r0::item:3"
    type="button"
  >
    3
  </button>
  <button
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black ds:font-bold"
    data-index="4"
    data-part="item"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r0::item:4"
    type="button"
  >
    4
  </button>
  <button
    aria-label="Next"
    class="ds:sm:min-w-[37px] ds:min-h-7 ds:min-w-7 ds:sm:min-h-[37px] ds:p-2 ds:rounded-full ds:flex ds:justify-center ds:items-center ds:cursor-pointer ds:bg-bg-gray-2 ds:text-black"
    data-part="next-trigger"
    data-scope="pagination"
    dir="ltr"
    id="pagination::r0::next"
    type="button"
  >
    <svg
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </button>
</nav>
`;
